src/familiar.ts

```diff
@@ -281,3 +363,149 @@ export function timeToMeatify(): boolean {
+
+type MarginalFamiliar = GeneralFamiliar & {
+  marginalValue: number;
+};
+
+let lepOutfitMeatPercent: number | null = null;
+let lepOutfitFamWeight: number | null = null;
+let nonLepOutfitMeatPercent: number | null = null;
+let nonLepOutfitFamWeight: number | null = null;
+let passiveSkillFamWeight = -1;
+
+export function setMarginalFamiliar(loc: Location, underwater: boolean): void {
+  let buffFamWeight = passiveSkillFamWeight < 0 ? 0 : passiveSkillFamWeight;
+  getActiveEffects().forEach((eff) => {
+    buffFamWeight += numericModifier(eff, "Familiar Weight");
+  });
+
+  if (passiveSkillFamWeight < 0) {
+    passiveSkillFamWeight = weightAdjustment() - buffFamWeight;
+    for (const slot of $slots`hat, back, shirt, weapon, off-hand, pants, acc1, acc2, acc3, familiar`) {
+      passiveSkillFamWeight -= numericModifier(equippedItem(slot), "Familiar Weight");
+    }
+    buffFamWeight += passiveSkillFamWeight;
+  }
+
+  // Compute optimal outfit for robort (rather than a generic lep) if we have one
+  if (have($familiar`Robortender`)) {
+    print("Computing leprechaun outfits for the first time...", "blue");
+    useFamiliar($familiar`Robortender`);
+    meatOutfit(false, undefined, false);
+    lepOutfitMeatPercent = 0;
+    lepOutfitFamWeight = 0;
+    for (const slot of $slots`hat, back, shirt, weapon, off-hand, pants, acc1, acc2, acc3, familiar`) {
+      lepOutfitMeatPercent += numericModifier(equippedItem(slot), "Meat Drop");
+      lepOutfitFamWeight += numericModifier(equippedItem(slot), "Familiar Weight");
+    }
+    print(`lepOutfitMeatPercent: ${lepOutfitMeatPercent}%`, "blue");
+    print(`lepOutfitFamWeight: ${lepOutfitFamWeight}%`, "blue");
+  }
+
+  const barf = loc === $location`Barf Mountain`;
+  const dropFamiliars = rotatingFamiliars
+    .map((fam): GeneralFamiliar => {
+      return {
+        familiar: fam.familiar,
+        expectedValue: garboValue(fam.drop) / (fam.expected[get(fam.pref)] ?? Infinity),
+        leprechaunMultiplier: findLeprechaunMultiplier(fam.familiar),
+      };
+    })
+    .concat(standardFamiliars())
+    .concat({
+      familiar: $familiar`Space Jellyfish`,
+      expectedValue: barf
+        ? garboValue($item`stench jelly`) /
+          (get("_spaceJellyfishDrops") < 5 ? get("_spaceJellyfishDrops") + 1 : 20)
+        : 0,
+      leprechaunMultiplier: 0,
+    })
+    .filter((fam) => have(fam.familiar) && (!underwater || fam.familiar.underwater))
+    .map((fam): MarginalFamiliar => {
+      const isLep = fam.leprechaunMultiplier > 0;
+      let additionalValue = 0;
+      if (isLep) {
+        if (!lepOutfitMeatPercent || !lepOutfitFamWeight) {
+          print("Computing leprechaun outfits for the first time...", "blue");
+          useFamiliar(fam.familiar);
+          meatOutfit(false, undefined, false);
+          lepOutfitMeatPercent = 0;
+          lepOutfitFamWeight = 0;
+          for (const slot of $slots`hat, back, shirt, weapon, off-hand, pants, acc1, acc2, acc3, familiar`) {
```

Comment:
yeah, I would prefer a `sum` to using a `for` loop here. A good rule of thumb is to try to avoid using `let` unless you're caching something.